client_auth.go

```diff
@@ -0,0 +1,122 @@
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/mendersoftware/log"
+	"github.com/pkg/errors"
+)
+
+type AuthGetter interface {
+	Request(identity, deviceID string) (interface{}, error)
+}
+
+// source of authorization data
+type AuthDataProvider interface {
+	// returns device public key
+	Pubkey() (string, error)
+	// returns tenant token
+	Token() (string, error)
+	// returns identity data
+	IdentityData() (interface{}, error)
+}
+
+type AuthClient struct {
+	client *http.Client
+}
+
+func NewAuthClient(conf httpsClientConfig) (*AuthClient, error) {
+	client, err := NewHttpClient(conf)
+	if client == nil {
+		return nil, errors.Wrap(err, "failed to create auth client HTTP client")
+	}
+
+	au := AuthClient{
+		client: client,
+	}
+	return &au, nil
+}
+
+func (u *AuthClient) Request(server string, datasrc AuthDataProvider) (interface{}, error) {
+
+	req, err := makeAuthRequest(server, datasrc)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to build auth request")
+	}
+
+	log.Debugf("making authorization request to server %s with req: %s", server, req)
+	rsp, err := u.client.Do(req)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to execute request")
+	}
+	defer rsp.Body.Close()
+
+	log.Debugf("receive response data")
+	data, err := ioutil.ReadAll(rsp.Body)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to receive response data")
+	}
+
+	// validate response
+	log.Debugf("received response data %v", data)
+
+	return data, nil
+}
+
+type authReqData struct {
+	IdData      interface{} `json:"id_data"`
+	TenantToken string      `json:"tenant_token"`
+	Pubkey      string      `json:"pubkey"`
+}
+
+func makeAuthRequest(server string, datasrc AuthDataProvider) (*http.Request, error) {
+	url := buildURL(server) + "/v0.1/auth_requests"
+
+	var err error
+	authd := authReqData{}
+
+	authd.IdData, err = datasrc.IdentityData()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to obtain identity data")
+	}
+
+	authd.TenantToken, err = datasrc.Token()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to obtain tenant token")
+	}
+
+	authd.Pubkey, err = datasrc.Pubkey()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to obtain device public key")
+	}
+
+	databuf := &bytes.Buffer{}
+	enc := json.NewEncoder(databuf)
+
+	err = enc.Encode(&authd)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to encode auth request")
+	}
+
+	req, err := http.NewRequest(http.MethodGet, url, databuf)
+	if err != nil {
+		return nil, err
```

Comment:
I like the way of returning error strings with `errors.Wrap()`. Here we can receive some weird error which might be hard to figure out later on.